### DISCUSSION TITLE
Using RCareWorld with ROS

### DISCUSSION THREAD

**Discussion by ALaws0n** [empriselab/RCareWorld#142] [Help - Using] [answered]
Hello,

I would like to try out RCareWorld with ROS. Do you have any examples or guides for configuring RCareWorld to work with ROS? 

Let me know if you have any questions.

**Comment by thisjustin123** [comment]
Hello! As far as I am aware, there is no current added support for using ROS in PhyRC RCareWorld.

I suppose then the fullest extent you can use ROS with at the moment is by writing a ROS connection to your python script/backend. From the python then you can directly interact with the platform.

**Reply by ALaws0n** [reply]
Hello, 

I read in the RCareWorld paper that "The Unity server communicates with ... the ROS client with Unity Robotics Hub..." and that "ROS makes it easier for users to perform motion planning with existing libraries including MoveIt! and OMPL". This is also mentioned in the RCareWorld presentation video around timestamp [7:15](https://youtu.be/mNy1cloWrP0?t=435). I am interested in using both of the aforementioned libraries. 

Are these features available?

**Reply by YoruCathy** [reply]
Hi @ALaws0n is it because you want to use ROS for the planning libraries? We do have it in [this](https://github.com/empriselab/RCareWorld/blob/phy-robo-care/pyrcareworld/pyrcareworld/demo/examples/example_ompl.py) example.

**Reply by ALaws0n** [reply]
Hi @YoruCathy 

That is correct. We are looking to use OMPL with RCareWorld. 

When we use the example_ompl.py script, we receive a continuous error message stating "An Error Occurred: OmplManagerAttr". The robot does complete its motion, but it clips through the obstacles in the environment. I believe this error is coming from '_parse_instance_data' in base_env.py. 

I attempted to troubleshoot this issue by passing in the OmplManagerAttr as an 'ext_attr'. This eliminates the error message but instead causes OMPL to find no solution. I have not yet attempted to mess with the start and target states in the example code. 

This issue is why I was looking into the aforementioned ROS support. Please advise on how I should proceed.

**Reply by YoruCathy** [reply]
I'll look into this issue and get back to you soon.

**Reply by thisjustin123** [reply]
Documented in #143 with a few observations

**Reply by ALaws0n** [reply]
Hello,

Thank you both for your help. I am adding a comment to the issue with more information. I hope this helps.

**Reply by YoruCathy** [reply]
Hi @ALaws0n , how did you install ompl on your end?

**Reply by ALaws0n** [reply]
Hi @YoruCathy , I installed OMPL from [source](https://github.com/ompl/ompl) and followed the instructions [here](https://github.com/lyfkyle/pybullet_ompl).

**Reply by YoruCathy** [reply]
Hi @ALaws0n , some quick updates: I'll be working on this on this PR https://github.com/empriselab/RCareWorld/pull/144
I'll post the updates there and let you know when I get a fix.

**Reply by YoruCathy** [reply]
Merged the PR https://github.com/empriselab/RCareWorld/pull/144

**Reply by ALaws0n** [reply]
Hello, 

Thank you for the update. I have a quick question. 

Does the change `executable_file="@editor"` mean that the Unity Editor must be running w/ the RCareWorld project open? 

Just want to make sure I understand the changes.

**Reply by thisjustin123** [reply]
Oops! Thanks for bringing that up. That shouldn't be `@editor`; it should be player_path. Fixing that now in #174.

Locally, feel free to just change that back to `player_path` if you want to run the example.

**Reply by YoruCathy** [reply, marked as answer]
Merged #174
